proposals/0019-graphql.md

```diff
@@ -76,7 +81,320 @@ The result of this query is a `GraphQLResult` that encodes to the following JSON
+
+A request string for this might be:
+
+```graphql
+mutation {
+  newPerson(name: "Tyler Endicott", age: 22, height: 1.63) {
+    name
+  }
+}
+```
+
+which would generate the response:
+
+```json
+{
+  "newPerson": {
+    "name": "Tyler Endicott"
+  }
+}
+```
+
+### Input Objects
+
+Sometimes we'd like to pass a complex argument. `Input`s allow us to do this and are declared by including an `Input` block in the API declaration, composed of `InputField`s. Our example can be changed to include a mutation that creates multiple new people, each passed as an input object:
+
+```swift
+struct NewPeopleArguments: Codable {
+    let individuals: [Person]
+}
+
+struct PersonResolver {
+    func newPeople(context: NoContext, arguments: NewPeopleArguments) -> [Person] {
+        return arguments.individuals.map { person in
+            Person(
+                name: person.name,
+                age: person.age,
+                height: person.height
+            )
+        }
+    }
+}
+
+struct PersonAPI : API {
+    typealias ContextType = NoContext
+    let resolver = PersonResolver()
+    let schema = try! Schema<PersonResolver, NoContext> {
+        Type(Person.self) {
+            Field("name", at: \.name)
+            Field("age", at: \.age)
+            Field("height", at: \.height)
+        }
+        Input(Person.self) {
+            InputField("name", at: \.name)
+            InputField("age", at: \.age)
+            InputField("height", at: \.height)
+        }
+        Mutation {
+            Field("newPeople", at: PersonResolver.newPeople) {
+                Argument("individuals", at: \.individuals)
+            }
+        }
+    }
+}
+```
+
+A request might look like:
+
+```graphql
+mutation {
+  newPeople(individuals: [
+    {name: "Tyler Endicott", age: 22, height: 1.63},
+    {name: "Angelo Pappas", age: 45, height: 1.91},
+  ]) {
+    name
+  }
+}
+```
+
+which would generate the response:
+
+```json
+{
+  "newPeople": [
+    {
+      "name": "Tyler Endicott",
+      "name": "Angelo Pappas"
+    }
+  ]
+}
+```
+
+### Subscriptions
+
+Subscriptions are reactive queries that return a result whenever an event occurs. This functionality is built on Swift Concurrency using `AsyncThrowingStream`. To create a subscription, include a `Subscription` block in the API declaration composed of `SubscriptionFields`. We can change our example API to include a subscription alert:
+
+```swift
+import Foundation
+import GraphQL
+
+struct PersonResolver {
+    func fiftyYearStormAlert(context: NoContext, arguments: NoArguments) -> ConcurrentEventStream<String> {
+        let asyncStream = AsyncThrowingStream<String, Error> { continuation in
+            let timer = Timer.scheduledTimer(
+                withTimeInterval: 60 * 60 * 24 * 365 * 50,
+                repeats: true
+            ) { _ in
+                continuation.yield("A 50-year storm is occurring!")
+            }
+        }
+        return ConcurrentEventStream<String>.init(asyncStream)
```

Comment:
small personal question not a problem for the incubation process, a bit weird name for this type -- is it just because there was some other EventStream already and you were looking to name one supporting swift concurrency?

Reply:
Yeah, I wrote the Subscription support a year and a half ago, back before Swift Concurrency. The canonical JavaScript implementation is based on an AsyncIterator, so as a workaround in Graphiti I abstracted out the event stream functionality into a protocol named `EventStream`, designed to be easily compatible with each reactive framework (`RxSwift`, `Combine`, etc).

When we added in Swift Concurrency support a few months ago, the `EventStream` name was already taken, and [Pioneer](https://github.com/d-exclaimation/pioneer/blob/main/Sources/Pioneer/Streaming/AsyncEventStream.swift) had already used `AsyncEventStream`, so in the interest of preserving the public API and reducing downstream headaches we went with with `ConcurrentEventStream`.

In a year or two, once we stop supporting pre-5.5 Swift and we've collected enough items for a new major release, I think we'll probably remove the protocol and rename `ConcurrentEventStream` to `EventStream`.

Reply:
I see, that sounds like a good plan there, thanks for the info!